pkg/sql/distsql_plan_bulk.go

```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
 	})
 	return planCtx, sqlInstanceIDs, nil
 }
+
+// PhysicalPlanMaker describes a function that makes a physical plan.
+type PhysicalPlanMaker func(context.Context, *DistSQLPlanner) (*PhysicalPlan, *PlanningCtx, error)
+
+// CalculatePlanGrowth returns the number of new or updated sql instances in one
+// physical plan relative to an old one, as well as that number as a fraction of
+// the number in the old one.
+func CalculatePlanGrowth(before, after *PhysicalPlan) (int, float64) {
+	var changed int
+	beforeSpecs := before.GenerateFlowSpecs()
+	afterSpecs := after.GenerateFlowSpecs()
+
+	// How many nodes are in after that are not in before, or are in both but
+	// have changed their spec?
+	for n, afterSpec := range afterSpecs {
+		if beforeSpec, ok := beforeSpecs[n]; !ok {
+			changed++ // not in before at all
+		} else if len(beforeSpec.Processors) != len(afterSpec.Processors) {
+			changed++ // in before with different procs
+		} else {
+			for i := range beforeSpec.Processors {
+				// TODO(dt): add equality check method gen, use it.
+				if afterSpec.Processors[i].Size() != beforeSpec.Processors[i].Size() {
+					changed++ // procs have differnet specs
+					break
+				}
+			}
+		}
+	}
+
+	return changed, float64(changed) / float64(len(beforeSpecs))
+}
+
+// ErrPlanChanged is a sentinel marker error for use to signal a plan changed.
+var ErrPlanChanged = errors.New("physical plan has changed")
+
+// PhysicalPlanChangeChecker returns a function which will periodically call the
+// passed function at the requested interval until the returned channel is
+// closed and compare the plan it returns to the passed initial plan, returning
+// and error if it has changed (as defined by CalculatePlanGrowth) by more than
+// the passed threshold. A threshold of 0 disables it.
+func PhysicalPlanChangeChecker(
+	ctx context.Context,
+	name string,
+	initial *PhysicalPlan,
+	fn PhysicalPlanMaker,
+	execCtx JobExecContext,
+	freq time.Duration,
```

Comment:
Would be nice to have frequency as a cluster setting.
If we want that to be caller controlled, we can pass `DurationSetting`

Reply:
no reason frequency can't be cluster setting (see usage in second commit). but it doesn't seem worth passing one in and making a *much* more complicated timer that adjusts on setting change (you'd need an on-change to wake it), so it's just up to the caller to call the setting and pass it.

Reply:
But this, a long running function.  It really isn't that much more difficult to reset the timer; but we almost always wish we could change a fixed frequency.

Reply:
It's pretty annoying/expensive to use the on-change API with a multiple waiters isn't it?

Reply:
Switched it to take a func() time.Duration